proposals/fit_curve.json

```diff
@@ -0,0 +1,83 @@
+{
+    "id": "fit_curve",
+    "summary": "Curve fitting",
+    "description": "Use non-linear least squares to fit a model function `y = f(x, parameters)` to data.",
+    "categories": [
+        "math"
+    ],
+    "experimental": true,
+    "parameters": [
+        {
+            "name": "data",
+            "description": "A labeled array of numbers. The array labels are the values for `x` (independent variable where the data is measured) and the array elements are the values for `y` (dependent data).",
```

Comment:
I wonder how easy it will be for a user to get the input data in this labeled array format.

Wouldn't it be easier to have just two input array parameters: `x` and `y`? I think that will feel more familiar for most users. And it;s more self-documenting.

Reply:
Also, the schema does not allow you to specify that the labels should be numeric, right?

Reply:
Indeed, good points. It's probably easier and more generic to support x and y separately, but this would basically require the implementation of array_labels in back-ends to be able to support this process.

Reply:
I've drafted a new proposal based on the discussion above. What do you think, @soxofaan ?

Reply:
In my opinion, the ideal case would be to just have a datacube as input: the back-end translates the time labels and takes care of them. The input datacube should have dimensions: time (which will be fitted to x in fit_curve) and bands (which will be fitted to y) and the fitting will be computed on all the available bands.

Reply:
@m-mohr  you use parameter names `data` for the dependent data (y) and `labels` for the independent variables (x).
Why not just `x` and `y`, which is standard terminology in machine learning literature and tools as far as I know.
Note that especially "labels" is tricky because that term is also often used as the target "y" you train with/for (in classification problems).

Reply:
> the ideal case would be to just have a datacube as input

That is also an interesting solution, and probably the shortest path for the user

> The input datacube should have dimensions: time (which will be fitted to x in fit_curve) and bands (which will be fitted to y) and the fitting will be computed on all the available bands.

you seem to have a particular use case in mind here: multiple timeseries to be fitted at the same time.
I wasn't thinking about more general use cases, not just  time based x'es, For example: fit EVI vs NDVI (so band vs band).

I'm not sure if we should directly aim for the most general solution (e.g. general `fit_curve`), or focus on just timeseries fitting (e.g. `fit_timeseries`).

Reply:
Well, yes, the specific use case is UC6, where we need to fit a seasonal function to each pixel in the data timeseries.

Reply:
> In my opinion, the ideal case would be to just have a datacube as input: the back-end translates the time labels and takes care of them.

If the only feasible way is to compute this on a temporal dimension, this could be a nice shortcut, but we should ensure that we don't end up with two similar processes, one defined on arrays and one defined on data cubes.

> Why not just `x` and `y`, which is standard terminology in machine learning literature and tools as far as I know.

In openEO we use x and y for numerical scalar values in mostly all (math related) processes. I'm not sure whether there are implementations that actually use these conventions for anything, but that may also be misleading for users then. FWIW, scipy uses ydata and xdata...

> Note that especially "labels" is tricky because that term is also often used as the target "y" you train with/for (in classification problems).

I've simply used this to make it work well with openEO terminology. In the normal workflow you'd get a labeled array in apply_dimension or so and then pass the "data" parameter to "data" and then call array_labels(data) and pass it to the labels parameter. That seems pretty intuitive (ignoring any background people may have).

> I'm not sure if we should directly aim for the most general solution (e.g. general `fit_curve`), or focus on just timeseries fitting (e.g. `fit_timeseries`).

Me neither, but we should ensure that we don't end up with two similar processes, one defined on arrays and one defined on data cubes. What other dimensions could you do curve fitting for?